### PR TITLE
Store scheduled events in DB as well

### LINKED ISSUE #157: Save required internal state to DB
## Problem Definition

The is some internal state that we need to persist, like `Context.scheduled_events`  and ` Context.waiting_transactions`. 

## Actual Outcome

Entries in these lists are lost after restarts.

## Expected Outcome

Items should still be available after a restart, with those deleted that do not make sense any more.

### PR DESCRIPTION
**Pull request by palango**
Merge after #159 .

Store scheduled events in DB as well

This commit moves scheduled events from memory to the database.
This makes them persistent, so that they are available after restarts
and makes sure that the DB is the single source of truth.

Fixes #157

### CODE REVIEW

File: src/monitoring_service/database.py [hunk 1]
```diff
@@ -143,6 +157,94 @@ def get_channel(self, token_network_address: str, channel_id: int) -> Optional[C
...
+    def upsert_scheduled_event(self, event: ScheduledEvent) -> None:
+        contained_event: SubEvent = cast(SubEvent, event.event)
+        values = [
+            hex(event.trigger_block_number),
+            EVENT_TYPE_ID_MAP[type(contained_event)],
+            contained_event.token_network_address,
+            hex(contained_event.channel_identifier),
+            contained_event.non_closing_participant,
+        ]
+        with self.conn:
```
Comment (karlb) [inline src/monitoring_service/database.py]: Transaction should be outside the Database class or you will lose the possibility to group different writes into the same transaction. Even grouping some reads along with a single write can be necessary for consistency when having concurrent db access.

Apparently, we want one transaction be write in many cases and the transaction handling is getting tedious. We could us autocommit on our connection by setting `isolation_level = None` when creating the connection. See https://docs.python.org/3.7/library/sqlite3.html#controlling-transactions.

File: src/monitoring_service/database.py [hunk 2]
```diff
@@ -13,11 +18,20 @@
...
 )
 
+SubEvent = Union[ActionMonitoringTriggeredEvent, ActionClaimRewardTriggeredEvent]
+
 log = structlog.get_logger(__name__)
 SCHEMA_FILENAME = os.path.join(
     os.path.dirname(os.path.realpath(__file__)),
     'schema.sql',
 )
+EVENT_ID_TYPE_MAP = {
```
Comment (karlb) [inline src/monitoring_service/database.py]: Maybe overkill for this simple case, but and `Enum` could be used here (along with `@unique`).

File: src/monitoring_service/database.py [hunk 3]
```diff
@@ -143,6 +157,94 @@ def get_channel(self, token_network_address: str, channel_id: int) -> Optional[C
...
+        rows = self.conn.execute(
+            """
+                SELECT * FROM scheduled_events
+                WHERE trigger_block_number <= ?
+            """,
+            [hex(max_trigger_block)],
+        ).fetchall()
+
+        def create_scheduled_event(row: sqlite3.Row) -> ScheduledEvent:
+            klass = EVENT_ID_TYPE_MAP[row['event_type']]
```
Comment (karlb) [inline src/monitoring_service/database.py]: The variable could be name `EventType`, which is more descriptive and gets around the spelling-mistake-to-avoid-keyword uglyness.

File: src/monitoring_service/database.py [hunk 4]
```diff
@@ -143,6 +157,94 @@ def get_channel(self, token_network_address: str, channel_id: int) -> Optional[C
...
+
+    def get_waiting_transactions(self) -> List[str]:
+        return [
+            row[0] for row in self.conn.execute(
+                "SELECT transaction_hash FROM waiting_transactions",
+            )
+        ]
+
+    def add_waiting_transaction(self, waiting_tx_hash: str) -> None:
+        with self.conn:
```
Comment (karlb) [inline src/monitoring_service/database.py]: Same as above.

File: src/monitoring_service/database.py [hunk 5]
```diff
@@ -143,6 +157,94 @@ def get_channel(self, token_network_address: str, channel_id: int) -> Optional[C
...
+            self.conn.execute(
+                "INSERT INTO waiting_transactions VALUES (?)",
+                [waiting_tx_hash],
+            )
+
+    def remove_waiting_transaction(self, tx_hash: str) -> None:
+        with self.conn:
```
Comment (karlb) [inline src/monitoring_service/database.py]: Same as above.

File: src/monitoring_service/schema.sql [hunk 6]
```diff
@@ -52,3 +52,18 @@ CREATE TABLE monitor_request (
...
+
+CREATE TABLE scheduled_events (
+    trigger_block_number    HEX_INT     NOT NULL,
+    event_type              INT NOT NULL CHECK (event_type >= 0 AND event_type <=1),
+
+    token_network_address   CHAR(42)    NOT NULL,
+    channel_identifier      HEX_INT     NOT NULL,
+    non_closing_participant CHAR(42)    NOT NULL,
+
+    PRIMARY KEY (trigger_block_number, channel_identifier, token_network_address)
```
Comment (karlb) [inline src/monitoring_service/schema.sql]: Can only a single event be scheduled for same block and channel?

File: src/monitoring_service/database.py (hunk 2 above)
Comment (palango) [reply src/monitoring_service/database.py]: I'd rather leave it for now. We can look into this again when we touch it.

**Comment by palango** [comment]
Updated PR to the feedback.

**Pull request merged**